File: tracker/src/main/java/com/cyrillrx/tracker/consumer/EventConsumer.java
package com.cyrillrx.tracker.consumer;

import com.cyrillrx.tracker.event.TrackEvent;

import java.util.Collection;

/**
 * @author Cyril Leroux
 *         Created on 20/04/16.
 */
public abstract class EventConsumer<EventContainer extends Collection<TrackEvent>>
        implements Runnable {

    protected static final TrackEvent STOP_EVENT = new TrackEvent.Builder()
            .setCategory("STOP_EVENT")
            .build();

    protected final EventContainer events;
    protected boolean running;

    public EventConsumer(EventContainer events) { this.events = events; }

    @Override
    public void run() {
        running = true;
        while (running) {
            consume();
        }
    }

    protected abstract void consume();
}

File: tracker/lib/src/androidMain/kotlin/com/cyrillrx/tracker/consumer/EventConsumer.kt
package com.cyrillrx.tracker.consumer

import com.cyrillrx.tracker.event.TrackEvent

abstract class EventConsumer<EventContainer : Collection<TrackEvent>>(
    protected val events: EventContainer
) : Runnable {

    protected var running: Boolean = false

    override fun run() {
        running = true
        while (running) {
            consume()
        }
    }

    protected abstract fun consume()

    class StopEvent : TrackEvent("STOP_EVENT")
}
